### PR TITLE
CORE-15586 Split verification from TransactionSignatureService into TransactionSignatureVerificationService

### PR DESCRIPTION
**Pull request by vlajos**
CORE-15586 Split verification from TransactionSignatureService into TransactionSignatureVerificationService to let it be injected separately in the verification sandbox. That way, we can keep the verification sandbox more streamlined.

### CODE REVIEW

File: components/ledger/ledger-common-flow/src/main/kotlin/net/corda/ledger/common/flow/impl/transaction/TransactionSignatureVerificationServiceImpl.kt [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+        publicKey: PublicKey
+    ) {
+        if (signatureWithMetadata.proof != null) {
+            transaction.confirmBatchMerkleSettingsMatch(signatureWithMetadata)
+            listOf(transaction).confirmHashPrefixesAreDifferent()
+        }
+        verifySignature(transaction.id, signatureWithMetadata, publicKey)
+    }
+
+    override fun verifySignature(
```
Comment (jmacmahonr3) [inline components/ledger/ledger-common-flow/src/main/kotlin/net/corda/ledger/common/flow/impl/transaction/TransactionSignatureVerificationServiceImpl.kt:69]: @vlajos would it be possible to take a list or a set of keys ? 

cc @szymonsztuka

**Review by jmacmahonr3 — CHANGES_REQUESTED** [review]

File: components/ledger/ledger-common-flow/src/main/kotlin/net/corda/ledger/common/flow/impl/transaction/TransactionSignatureVerificationServiceImpl.kt (hunk 1 above)
Comment (vlajos) [reply components/ledger/ledger-common-flow/src/main/kotlin/net/corda/ledger/common/flow/impl/transaction/TransactionSignatureVerificationServiceImpl.kt:69]: @jmacmahonr3 @szymonsztuka 
Generally, we prefer the most straightforward approach in APIs and let the consumers build the more complicated scenarios from the building blocks.
In this case, I think the caller side could use a similar function:
```
fun verifySignature(
    secureHash: SecureHash,
    signatureWithMetadata: DigitalSignatureAndMetadata,
    publicKeys: List<PublicKey>,
    digestService: DigestService
) {
    val digestAlgorithmOfKeyId = signatureWithMetadata.signature.by.algorithm
    val publicKey = requireNotNull(publicKeys.firstOrNull {
        digestService.hash(
            it.encoded,
            DigestAlgorithmName(digestAlgorithmOfKeyId)
        ) == signatureWithMetadata.signature.by
    }) {
        "The key Id of the provided signature does not match the provided public keys' id."
    }
    transactionSignatureVerificationService.verifySignature(secureHash, signatureWithMetadata, publicKey)
}
```
Could this work for you?

File: components/ledger/ledger-common-flow/src/main/kotlin/net/corda/ledger/common/flow/impl/transaction/TransactionSignatureVerificationServiceImpl.kt [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+            checkSignatureSpec(
+                signatureWithMetadata.metadata.signatureSpec,
+                publicKey
+            )
+
+        val proof = signatureWithMetadata.proof
+        val signedHash = if (proof == null) {   // Simple signature
+            secureHash
+        } else {                                // Batch signature
+            require(proof.leaves.filter { secureHash.bytes contentEquals it.leafData }.size == 1) {
```
Comment (szymonsztuka) [inline components/ledger/ledger-common-flow/src/main/kotlin/net/corda/ledger/common/flow/impl/transaction/TransactionSignatureVerificationServiceImpl.kt:89]: So essentially for batch signature (which is returned by the signing "activity") we check it contains a given  transation id. A question not necessarily related to this code, there is no semantics here what "transaction id" is, but we know in Corda UTXO model this should be a hash value of the transaction (content/binary)?

File: components/ledger/ledger-common-flow/src/main/kotlin/net/corda/ledger/common/flow/impl/transaction/TransactionSignatureVerificationServiceImpl.kt (hunk 1 above)
Comment (szymonsztuka) [reply components/ledger/ledger-common-flow/src/main/kotlin/net/corda/ledger/common/flow/impl/transaction/TransactionSignatureVerificationServiceImpl.kt:69]: Would DigestService be available in a class extending `net.corda.v5.ledger.utxo.Contract` ? I think yes, but haven't tried this:
```
class NotaryVerifyContract : Contract {
    @CordaInject
    lateinit var transactionSignatureVerificationService: TransactionSignatureVerificationService
    @CordaInject
    lateinit var digestService: DigestService

```

File: testing/cpbs/ledger-utxo-demo-contract/src/main/kotlin/com/r3/corda/demo/utxo/contract/notaryverify/NotaryVerifyContract.kt [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+import net.corda.v5.application.flows.CordaInject
+import net.corda.v5.ledger.common.transaction.TransactionSignatureVerificationService
+import net.corda.v5.ledger.utxo.Contract
+import net.corda.v5.ledger.utxo.ContractState
+import net.corda.v5.ledger.utxo.VisibilityChecker
+import net.corda.v5.ledger.utxo.transaction.UtxoLedgerTransaction
+
+class NotaryVerifyContract : Contract {
+    @CordaInject
+    lateinit var transactionSignatureVerificationService: TransactionSignatureVerificationService
```
Comment (szymonsztuka) [inline testing/cpbs/ledger-utxo-demo-contract/src/main/kotlin/com/r3/corda/demo/utxo/contract/notaryverify/NotaryVerifyContract.kt]: So just to clarify, I can use this service inside Contract verification code like here below?

**Review by szymonsztuka — DISMISSED** [review]

File: components/ledger/ledger-common-flow/src/main/kotlin/net/corda/ledger/common/flow/impl/transaction/TransactionSignatureVerificationServiceImpl.kt (hunk 1 above)
Comment (vlajos) [reply components/ledger/ledger-common-flow/src/main/kotlin/net/corda/ledger/common/flow/impl/transaction/TransactionSignatureVerificationServiceImpl.kt:69]: Yes. There is a related e2e test CorDapp with its test: https://github.com/corda/corda-e2e-tests/blob/release/5.1/apps/ledger-utxo-demo-contract/src/main/kotlin/com/r3/corda/demo/utxo/contract/ValidVerifyInjectionContract.kt

File: components/ledger/ledger-common-flow/src/main/kotlin/net/corda/ledger/common/flow/impl/transaction/TransactionSignatureVerificationServiceImpl.kt (hunk 2 above)
Comment (vlajos) [reply components/ledger/ledger-common-flow/src/main/kotlin/net/corda/ledger/common/flow/impl/transaction/TransactionSignatureVerificationServiceImpl.kt:89]: `TransactionSignature...ServiceImpl.kt` does not care too much about the transaction ids apart from expecting `SecureHash`s. (Wrapped in `TransactionWithMetadata` in some cases.)
In platform code, these are transaction ids calculated as the root hash of the transaction's Merkle tree. But this service does not check anything like this.

File: testing/cpbs/ledger-utxo-demo-contract/src/main/kotlin/com/r3/corda/demo/utxo/contract/notaryverify/NotaryVerifyContract.kt (hunk 3 above)
Comment (vlajos) [reply testing/cpbs/ledger-utxo-demo-contract/src/main/kotlin/com/r3/corda/demo/utxo/contract/notaryverify/NotaryVerifyContract.kt]: Yes. https://github.com/corda/corda-runtime-os/pull/4445#discussion_r1297417039

File: applications/workers/workers-smoketest/src/smokeTest/kotlin/net/corda/applications/workers/smoketest/ledger/UtxoLedgerTests.kt [hunk 4]
```diff
@@ -169,6 +169,41 @@ class UtxoLedgerTests {
         assertThat(parsedPeekFlowResult.outputs).singleElement().extracting { it.testField }.isEqualTo(evolvedMessage)
     }
 
+    @Test
```
Comment (blsemo) [inline applications/workers/workers-smoketest/src/smokeTest/kotlin/net/corda/applications/workers/smoketest/ledger/UtxoLedgerTests.kt]: Are you planning to move this test to the e2e test repo before merging? I don't think we should add more smoke tests...

**Review by blsemo — COMMENTED** [review]
Looks fine to me, apart from introducing a new smoke test in the runtime-os repo.

File: applications/workers/workers-smoketest/src/smokeTest/kotlin/net/corda/applications/workers/smoketest/ledger/UtxoLedgerTests.kt (hunk 4 above)
Comment (vlajos) [reply applications/workers/workers-smoketest/src/smokeTest/kotlin/net/corda/applications/workers/smoketest/ledger/UtxoLedgerTests.kt]: Yes. That is the plan. (I mentioned it on slack too.)

File: components/ledger/ledger-common-flow/src/main/kotlin/net/corda/ledger/common/flow/impl/transaction/TransactionSignatureVerificationServiceImpl.kt (hunk 1 above)
Comment (szymonsztuka) [reply components/ledger/ledger-common-flow/src/main/kotlin/net/corda/ledger/common/flow/impl/transaction/TransactionSignatureVerificationServiceImpl.kt:69]: cool!

File: testing/cpbs/ledger-utxo-demo-contract/src/main/kotlin/com/r3/corda/demo/utxo/contract/notaryverify/NotaryVerifyContract.kt (hunk 3 above)
Comment (szymonsztuka) [reply testing/cpbs/ledger-utxo-demo-contract/src/main/kotlin/com/r3/corda/demo/utxo/contract/notaryverify/NotaryVerifyContract.kt]: perfect!

**Review by blsemo — DISMISSED** [review]
LGTM

**Review by jmacmahonr3 — DISMISSED** [review]

**Review by blsemo — APPROVED** [review]
LGTM

**Pull request merged**